### DISCUSSION TITLE
How to expire model using AsyncSession?

### DISCUSSION THREAD

**Discussion by uselessscat** [sqlalchemy/sqlalchemy#9868] [Usage Questions] [answered]
Im trying to expire a model in a test case:

```python
async def test_example(
    self,
    db_session,
    authenticated_client,
):
    client, _, user = authenticated_client

    # client.post(...)

    db_session.expire(user)

    assert user.name == "test"
```

But this gives the following error:

>\> assert user.name == "test"
../local/lib/python3.11/site-packages/sqlalchemy/orm/attributes.py:563: in __get__
    return self.impl.get(state, dict_)  # type: ignore[no-any-return]
../local/lib/python3.11/site-packages/sqlalchemy/orm/attributes.py:1084: in get
    value = self._fire_loader_callables(state, key, passive)
../local/lib/python3.11/site-packages/sqlalchemy/orm/attributes.py:1114: in _fire_loader_callables
    return state._load_expired(state, passive)
../local/lib/python3.11/site-packages/sqlalchemy/orm/state.py:798: in _load_expired
    self.manager.expired_attribute_loader(self, toload, passive)
../local/lib/python3.11/site-packages/sqlalchemy/orm/loading.py:1626: in load_scalar_attributes
    result = load_on_ident(
../local/lib/python3.11/site-packages/sqlalchemy/orm/loading.py:482: in load_on_ident
    return load_on_pk_identity(
../local/lib/python3.11/site-packages/sqlalchemy/orm/loading.py:668: in load_on_pk_identity
    session.execute(
../local/lib/python3.11/site-packages/sqlalchemy/orm/session.py:2232: in execute
    return self._execute_internal(
../local/lib/python3.11/site-packages/sqlalchemy/orm/session.py:2127: in _execute_internal
    result: Result[Any] = compile_state_cls.orm_execute_statement(
../local/lib/python3.11/site-packages/sqlalchemy/orm/context.py:293: in orm_execute_statement
    result = conn.execute(
../local/lib/python3.11/site-packages/sqlalchemy/engine/base.py:1413: in execute
    return meth(
../local/lib/python3.11/site-packages/sqlalchemy/sql/elements.py:483: in _execute_on_connection
    return connection._execute_clauseelement(
../local/lib/python3.11/site-packages/sqlalchemy/engine/base.py:1637: in _execute_clauseelement
    ret = self._execute_context(
../local/lib/python3.11/site-packages/sqlalchemy/engine/base.py:1846: in _execute_context
    return self._exec_single_context(
../local/lib/python3.11/site-packages/sqlalchemy/engine/base.py:1987: in _exec_single_context
    self._handle_dbapi_exception(
../local/lib/python3.11/site-packages/sqlalchemy/engine/base.py:2347: in _handle_dbapi_exception
    raise exc_info[1].with_traceback(exc_info[2])
../local/lib/python3.11/site-packages/sqlalchemy/engine/base.py:1968: in _exec_single_context
    self.dialect.do_execute(
../local/lib/python3.11/site-packages/sqlalchemy/engine/default.py:920: in do_execute
    cursor.execute(statement, parameters)
../local/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py:563: in execute
    self._adapt_connection.await_(
../local/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py:117: MissingGreenlet

> sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can't call await_only() here. Was IO attempted in an unexpected place? (Background on this error at: https://sqlalche.me/e/20/xd2s)

Also if try with async:

```python
await db_session.expire(user)
```

I got:
>\> await db_session.expire(user)
TypeError: object NoneType can't be used in 'await' expression

AsyncSession.expire and AsyncSession.expire_all gives the same error but AsyncSession.refresh does not:

This code works as expected:

```python
async def test_example(
    self,
    db_session,
    authenticated_client,
):
    client, _, user = authenticated_client

    # client.post(...)

    await db_session.refresh(user)

    assert user.name == "Norma Fisher" # PASSED
```

The fixture db_session looks:

```python
from sqlalchemy.ext.asyncio import (
    AsyncEngine,
    AsyncSession,
    async_sessionmaker,
    create_async_engine,
)

@pytest.fixture(scope="session")
async def db_session(settings):
    engine: AsyncEngine = create_async_engine(
        settings.database_uri,
        pool_pre_ping=True,
    )

    session_maker: async_sessionmaker[AsyncSession] = async_sessionmaker(
        autocommit=False,
        autoflush=False,
        expire_on_commit=False,
        bind=engine,
    )

    async with session_maker() as session:
        yield session
```

Im using sqlalchemy 2.0.15 and asyncpg 0.27.0

How I can do an expire to invalidate model fields?

**Comment by zzzeek** [comment, marked as answer]
the expire is working as expected, the error you show is the next line when you attempt to access `user.name`, which also fails as would be expected since you [can't do straight lazy loading with asyncio](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#preventing-implicit-io-when-using-asyncsession).

to access expired fields under asyncio, use the [AsyncAttrs](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncAttrs) mixin on your class:

```py

from sqlalchemy.ext.asyncio import AsyncAttrs

class MyClass(AsyncAttrs, Base):
    # ...

```

then access ``user.name`` after expiration as:

```py

assert (await user.async_attrs.name) == "test"
```

**Reply by uselessscat** [reply]
Thanks for the explanation and the fast response, now it works.

But another question, Im using map_imperatively:

```python
# domain/entities.py
@dataclass
class User:
    id: int
    name: str
```

```python
# models.py
from sqlalchemy.orm import registry

default_registry = registry()

user_table = Table(
    "users",
    default_registry.metadata,

    Column("id", Integer, primary_key=True, autoincrement=True),
    Column("name", String(100)),
)

default_registry.map_imperatively(User, user_table)
```

How can I add the AsyncAttrs to the model without intervening the dataclass?

**Reply by zzzeek** [reply]
I was just expressing this same opinion on #9869, another user trying to do the exact same kind of design (oddly enough, at the same time...) where I dont really think it's appropriate from a "correctness" pov to use imperative mappings at all, especially with dataclasses, so for me that makes the problem much easier which is to use the [Declarative Dataclass Mapping](https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#declarative-dataclass-mapping) feature instead.

if you dont want to intervene with the dataclass at all then you can't use `AsyncAttrs`.    you would need to either use [all the other ways of avoiding lazy loading](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#preventing-implicit-io-when-using-asyncsession) (see all the additional bullets)  or you can port the internals of `AsyncAttrs` into a standalone function:

```py
from sqlalchemy.util import greenlet_spawn

assert (await greenlet_spawn(lambda: user.name) == "test"
```
